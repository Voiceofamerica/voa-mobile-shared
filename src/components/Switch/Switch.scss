
$size: 10px;
$toggleSize: 20px;
$ratio: 3;
$toggleOffset: ($toggleSize - $size) / 2;
$toggleOnOffset: ($size * $ratio - $toggleOffset - $size);

.innerSwitch {
  position: relative;
  width: $size * $ratio + 2;
  height: $size + 2;
  border-radius: $size / 2;
  background: #fff;
  border: 1px solid #ddd;
  transition: background 0.3s;

  .toggle {
    height: $toggleSize;
    width: $toggleSize;
    position: absolute;
    left: -$toggleOffset;
    top: -$toggleOffset;
    border-radius: $toggleSize;
    transition: left 0.3s;
    background: #ddd;
    border: 1px solid #bbb;
  }

  &.on {
    background: #ada;
    .toggle {
      left: $toggleOnOffset;
    }
  }
}
