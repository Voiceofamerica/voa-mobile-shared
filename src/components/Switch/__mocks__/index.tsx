
import * as React from 'react'

export default (props) => (
  React.createElement('switch', { props }, props.children)
)
