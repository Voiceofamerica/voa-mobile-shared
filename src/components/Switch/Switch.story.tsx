
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Switch from './Switch'

class ToggleableSwitch extends React.Component<{}, { on: boolean }> {
  state = {
    on: false,
  }

  render () {
    return <Switch value={this.state.on} onClick={on => this.setState({ on })} />
  }
}

storiesOf('Switch', module)
  .add('display off', () => (
    <Switch value={false} />
  ))
  .add('display on', () => (
    <Switch value={true} />
  ))
  .add('interactible', () => (
    <ToggleableSwitch />
  ))
