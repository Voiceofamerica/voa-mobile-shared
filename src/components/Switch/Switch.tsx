
import * as React from 'react'

import { innerSwitch, toggle, on } from './Switch.scss'

export interface Props {
  value: boolean
  onClick?: (newVal: boolean) => void
  className?: string
  style?: React.CSSProperties
}

class Ticket extends React.Component<Props> {
  render () {
    const { value = false, onClick = () => null, className = '', style = {} } = this.props
    const baseClassName = value ? `${innerSwitch} ${on}` : innerSwitch

    return (
      <div className={`${baseClassName} ${className}`} onClick={() => onClick(!value)} style={style}>
        <div className={toggle} />
      </div>
    )
  }
}

export default Ticket
