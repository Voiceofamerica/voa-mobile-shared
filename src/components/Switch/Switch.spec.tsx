
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import Switch from './Switch'

describe(`<${Switch.name} />`, () => {
  describe('shapshots', () => {
    it('should render toggled off', () => {
      const element = create((
        <Switch value={false} />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render toggled on', () => {
      const element = create((
        <Switch value={true} />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given className', () => {
      const element = create((
        <Switch value={false} className='myClass' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given style', () => {
      const element = create((
        <Switch value={false} style={{ top: 0, left: 0, right: 0, bottom: 0 }} />
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('onClick', () => {
    it('should be called whenever the element is clicked', () => {
      const onClick = jest.fn()

      const item = shallow((
        <Switch value={false} onClick={onClick} />
      ))

      item.simulate('click')

      expect(onClick).toHaveBeenCalledTimes(1)
    })

    it('should not throw an error if not provided and the element is clicked', () => {
      const item = shallow((
        <Switch value={false} />
      ))

      item.simulate('click')
    })
  })
})
