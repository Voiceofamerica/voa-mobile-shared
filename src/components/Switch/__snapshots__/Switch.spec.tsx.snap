// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ticket /> shapshots should render toggled off 1`] = `
<div
  className="innerSwitch "
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="toggle"
  />
</div>
`;

exports[`<Ticket /> shapshots should render toggled on 1`] = `
<div
  className="innerSwitch on "
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="toggle"
  />
</div>
`;

exports[`<Ticket /> shapshots should render with the given className 1`] = `
<div
  className="innerSwitch myClass"
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="toggle"
  />
</div>
`;

exports[`<Ticket /> shapshots should render with the given style 1`] = `
<div
  className="innerSwitch "
  onClick={[Function]}
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
>
  <div
    className="toggle"
  />
</div>
`;
